### PR TITLE
feat: gate WMMA GEMM dispatch behind env

### PR DESCRIPTION
**Pull request by donghanbj**
# feat(wmma): full ONNX Gemm support in WMMA GEMM kernel

## Summary

Extend `hip_gemm_wmma_fp16` from a bare `C = A * B` kernel into a standard ONNX
Gemm entry point:

```
output[M,N] = alpha * op(A) * op(B) + beta * C
```

The kernel now covers the common ONNX Gemm / `nn.Linear` shapes directly on the
RDNA 3+ WMMA path, while transparently falling back to hipBLASLt for any
configuration it does not implement.

## Motivation

The previous WMMA kernel only computed `C = A * B` (no transpose, no bias, no
scaling), so the `wrap_gemm` dispatch had to gate on a very narrow case
(`transA == 0 && transB == 0 && C == nullptr && alpha == 1`). Most real Gemm
nodes — especially `nn.Linear` (which ships its weight as `transB = 1`) and
nodes with a row-broadcast bias — never qualified and always fell through to
hipBLASLt. This change makes the WMMA path applicable to those nodes and lets
all the configuration validation live next to the kernel.

### `3rd-party/custom_kernels/hip/gemm_wmma_kernel.hip`
- New template parameter `bool TransB` selecting B layout:
  - `TransB == false`: B is `[K, N]` row-major (ONNX `transB = 0`).
  - `TransB == true`: B is `[N, K]` row-major (ONNX `transB = 1`, `nn.Linear`
    weight) — the K elements of a column are contiguous, so B is vector-loaded.
- Vectorized fragment loads: A is loaded as a single 32-byte `half16` per row
  instead of 16 scalar loads (B too when `TransB == 1`). The `transB == 0` path
  keeps the strided, already-coalesced scalar B load.
- Optional row-broadcast bias (`[N]` / `[1, N]`) folded into the output write.
- Host entry rewritten to the full ONNX Gemm signature and now performs
  capability + alignment validation:
  - returns `> 0` ("not handled, fall back") for unsupported configs,
  - returns `< 0` on launch error,
  - returns `0` when computed by WMMA.
- 32-byte base-pointer alignment guard for the vectorized loads (A always; B
  when `transB == 1`) — a misaligned base falls back instead of faulting.

### `3rd-party/custom_kernels/include/hip_custom_kernels.h`
- Updated the `hip_gemm_wmma_fp16` declaration and documentation to the new
  ONNX Gemm signature and the `> 0 / 0 / < 0` return contract.

### `lib/Runtime/real/gemm.cpp`
- `wrap_gemm` dispatch relaxed to gate only on `fp16 && K % 16 == 0 && N % 16 == 0`;
  all remaining ONNX Gemm attributes (transA / transB / alpha / beta / shape of
  C) are validated inside the kernel host function.
- `rc > 0` now falls through to hipBLASLt silently; `rc < 0` logs a launch error
  before falling back.

## Supported configurations (WMMA path)

| Attribute | Supported |
|-----------|-----------|
| dtype     | fp16 |
| `K`, `N`  | multiples of 16 |
| `transA`  | `0` only |
| `transB`  | `0` (B `[K,N]`) or `1` (B `[N,K]`) |
| `alpha`   | `1.0` |
| `beta`    | `1.0` |
| `C`       | absent, or row-broadcast bias `[N]` / `[1,N]` (`cDim0 == 1 && cDim1 == N`) |
| base ptr  | A 32-byte aligned (and B when `transB == 1`) |

Anything outside this set returns a positive code and `wrap_gemm` falls back to
hipBLASLt — behavior is always correct, only the fast path is conditional.

## HIPDNN_EP_GEMM_WMMA=1
- The WMMA path is opt-in via `HIPDNN_EP_GEMM_WMMA=1` (default: hipBLASLt).
- With `HIPDNN_EP_DEBUG=1`, a dispatched call logs
  `[REAL] wrap_gemm: dispatch=WMMA ...`.
- This is a custom-kernel change: after rebuilding, delete stale compiled model
  DLLs (`del %TEMP%\morphizen_mlir_*`) so the new bitcode is picked up.
- Validate correctness on the gfx1151 host (numeric tests), especially the
  `transB = 1` weight path that exercises the vectorized B load.

## Qwen3.5-9B-rtn-int4-int8-128gs-fp16-onnx-gpu_space\vision\single_op

Test models live under this path. The 6 Gemm folders exercised here:
`Gemm_linear`, `Gemm_linear_1`, `Gemm_linear_2`, `Gemm_linear_3`,
`Gemm_linear_108`, `Gemm_linear_109` (each with a `*_np1024.onnx` variant used
below). All are fp16, `transA=0`, `transB=1`, `alpha=beta=1`, with an `[N]` bias.

## Performance — np1024 variants (real workload, M = 1024/256)

WMMA=1 (this change) vs WMMA=0 (hipBLASLt). Throughput in inferences/sec
(higher is better), `-t 60 -c 1 -s`, fp16, all `transB=1` + `[N]` bias.
dynamic 
| Model (M×K×N) | hipBLASLt (WMMA=0) | WMMA=1 | Speedup |
|---|---|---|---|
Gemm_linear | 57.33 | 64.30 | +12.2%
Gemm_linear_1 | 57.56 | 62.03 | +7.8%
Gemm_linear_2 | 57.68 | 67.23 | +16.6%
Gemm_linear_3 | 56.22 | 62.60 | +11.3%
Gemm_linear_108 | 56.54 | 61.94 | +9.6%
Gemm_linear_109 | 61.53 | 61.95 | +0.7%

## Files

```
3rd-party/custom_kernels/hip/gemm_wmma_kernel.hip  | 136 +++++++++++++++---
3rd-party/custom_kernels/include/hip_custom_kernels.h |  45 ++++--
lib/Runtime/real/gemm.cpp                          |  38 +++---
3 files changed, 178 insertions(+), 41 deletions(-)
```
## Qwen3.5-9B Result

Model | Case | Performance |   |   |   |   |   |   | Accuracy |  
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
  |   | TTFT (s) |   | TPS |   | CPU Peak Memory (GB) | GPU Peak Memory (GB) | PPL |  
  |   | Baseline | Current | Baseline | Current | Baseline | Current | Baseline | Current | Baseline | Current
Qwen3.5-9B-fp16-ve-fp16-int4-text-gs32-dml | Prompt 2048 / Output 64 / genai_config_MorphiZenEP | 8.7115 | 8.8331(1.4%) | 26.92 | 26.59(-1.22%) | - | - | 18.6104 | 18.5544(-0.3%) |  

http://xsjnts.xilinx.com/ROCm/compare?baseline_job_name=daily_hipdnn_ep_oga&baseline_build_id=&your_job_name=daily_hipdnn_ep_oga&your_build_id=&base_url=http%3A%2F%2Fxsjnts.xilinx.com%2FROCm%2Fhipdnn-detail%3Fdatabase%3Ddevelopment%26collection%3Ddaily_hipdnn_ep_oga%26id%3D6a336aa3f427aacfec05a8cf&curr_url=http%3A%2F%2Fxsjnts.xilinx.com%2FROCm%2Fhipdnn-detail%3Fdatabase%3Ddevelopment%26collection%3Ddaily_hipdnn_ep_oga%26id%3D6a33b7ee6f8fae877a9b2eea

## Qwen3.5-35B Result

Model | Case | Performance |   |   |   |   |   |   | Accuracy
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
  |   | TTFT (s) |   | TPS |   | CPU Peak Memory (GB) | GPU Peak Memory (GB) | PPL
  |   | Baseline | Current | Baseline | Current | Baseline | Current | Baseline | Current | Baseline
Qwen3.5-35B-A3B-fp16-ve-fp16-int4-text-gs32-dml | Prompt 19 / Output 128 / genai_config_AMDGPU | 10.0933 | 10.0486(-0.44%) | 36.31 | 36.04(-0.74%) | 3.2668 | 3.2471(-0.6%) | 33.268 | 33.2083(-0.18%)
Qwen3.5-35B-A3B-fp16-ve-fp16-int4-text-gs32-dml | Prompt 128 / Output 128 / genai_config_AMDGPU | 10.4652 | 10.5123(0.45%) | 35.72 | 34.57(-3.23%) | 3.444 | 3.424(-0.58%) | 33.6613 | 33.6026(-0.17%)
Qwen3.5-35B-A3B-fp16-ve-fp16-int4-text-gs32-dml | Prompt 1400 / Output 128 / genai_config_AMDGPU | 17.7001 | 17.9126(1.2%) | 35.36 | 35.02(-0.97%) | 6.2434 | 6.1726(-1.13%) | 38.1267 | 38.0632(-0.17%)
Qwen3.5-35B-A3B-fp16-ve-fp16-int4-text-gs32-dml | Prompt 2048 / Output 128 / genai_config_AMDGPU | 21.5875 | 20.2741(-6.08%) | 33.46 | 35.1(4.92%) | 8.2449 | 8.184(-0.74%) | 40.144 | 40.0849(-0.15%)

http://xsjnts.xilinx.com/ROCm/compare?base_url=http%3A%2F%2Fxsjnts.xilinx.com%2FROCm%2Fhipdnn-detail%3Fdatabase%3Ddevelopment%26collection%3Ddaily_hipdnn_ep_oga%26id%3D6a3416a96f8fae877a9b3260&base_job=&base=&curr_url=http%3A%2F%2Fxsjnts.xilinx.com%2FROCm%2Fhipdnn-detail%3Fdatabase%3Ddevelopment%26collection%3Ddaily_hipdnn_ep_oga%26id%3D6a341be46f8fae877a9b326d&curr_job=&curr=

### CODE REVIEW

**Comment by amd-yanjunz** [comment]
/run-test models=Qwen3.5-9B-rtn-int4-int8-128gs-fp16-onnx-gpu_text

**Comment by amd-yanjunz** [comment]
ROCm OGA CI finished with result: UNSTABLE.

- PR: ROCm/onnx-hipdnn-ep#257
- Source comment: [models=Qwen3.5-9B-rtn-int4-int8-128gs-fp16-onnx-gpu_text](https://github.com/ROCm/onnx-hipdnn-ep/pull/257#issuecomment-4668308127)
- TEST_MODE: both
- MODELZOO: LLM_ROCm_release
- TEST_MODELS: Qwen3.5-9B-rtn-int4-int8-128gs-fp16-onnx-gpu_text
- Jenkins build: http://xsjvitisaijenkins:8080/job/daily_hipdnn_ep_oga/655/
- NTS report: http://xsjnts.xilinx.com/ROCm/hipdnn-detail?database=development&collection=daily_hipdnn_ep_oga&id=6a292f08aacbafe0971a1d31

**Comment by donghanbj** [comment]
/run-test models=Qwen3.5-9B-rtn-int4-int8-128gs-fp16-onnx-gpu_text

**Comment by amd-yanjunz** [comment]
ROCm OGA CI finished with result: UNSTABLE.

- PR: ROCm/onnx-hipdnn-ep#257
- Source comment: [models=Qwen3.5-9B-rtn-int4-int8-128gs-fp16-onnx-gpu_text](https://github.com/ROCm/onnx-hipdnn-ep/pull/257#issuecomment-4703095339)
- TEST_MODE: both
- MODELZOO: LLM_ROCm_release
- TEST_MODELS: Qwen3.5-9B-rtn-int4-int8-128gs-fp16-onnx-gpu_text
- Jenkins build: http://xsjvitisaijenkins:8080/job/daily_hipdnn_ep_oga/726/
- NTS report: http://xsjnts.xilinx.com/ROCm/hipdnn-detail?database=development&collection=daily_hipdnn_ep_oga&id=6a2f1c077243c6ea68e090a3

**Comment by amd-yanjunz** [comment]
ROCm OGA CI finished with result: UNSTABLE.

- PR: ROCm/onnx-hipdnn-ep#257
- Source comment: [models=Qwen3.5-9B-fp16-ve-fp16-int4-text-gs32-dml](https://github.com/ROCm/onnx-hipdnn-ep/pull/257#issuecomment-4705687652)
- TEST_MODE: both
- MODELZOO: LLM_ROCm_release
- TEST_MODELS: Qwen3.5-9B-fp16-ve-fp16-int4-text-gs32-dml
- Jenkins build: http://xsjvitisaijenkins:8080/job/daily_hipdnn_ep_oga/732/

**Comment by amd-yanjunz** [comment]
ROCm OGA CI finished with result: UNSTABLE.

- PR: ROCm/onnx-hipdnn-ep#257
- Source comment: [models=qwen35_9b](https://github.com/ROCm/onnx-hipdnn-ep/pull/257#issuecomment-4705809255)
- TEST_MODE: both
- MODELZOO: LLM_ROCm_dynamic
- TEST_MODELS: Qwen3.5-9B-fp16-ve-fp16-int4-text-gs32-dml
- Jenkins build: http://xsjvitisaijenkins:8080/job/daily_hipdnn_ep_oga/734/
- NTS report: http://xsjnts.xilinx.com/ROCm/hipdnn-detail?database=development&collection=daily_hipdnn_ep_oga&id=6a2fb2fda86f38dff415a540

**Comment by amd-yanjunz** [comment]
ROCm OGA CI finished with result: ABORTED.

- PR: ROCm/onnx-hipdnn-ep#257
- Source comment: [models=qwen35_9b](https://github.com/ROCm/onnx-hipdnn-ep/pull/257#issuecomment-4719900564)
- TEST_MODE: both
- MODELZOO: LLM_ROCm_dynamic
- TEST_MODELS: Qwen3.5-9B-fp16-ve-fp16-int4-text-gs32-dml
- Jenkins build: http://xsjvitisaijenkins:8080/job/daily_hipdnn_ep_oga/792/
